# 2. According to Second Substitute Bill BL2018-1202 (as amended) (https://web.archive.org/web/20181019234657/https://www.nashville.gov/Metro-Clerk/Legislative/Ordinances/Details/7d2cf076-b12c-4645-a118-b530577c5ee8/2015-2019/BL2018-1202.aspx), all permitted operators will first clean data before providing or reporting data to Metro. Data processing and cleaning shall include:  
## * Removal of staff servicing and test trips  
## * Removal of trips below one minute  
## * Trip lengths are capped at 24 hours  
## Are the scooter companies in compliance with the second and third part of this rule? 

### ------------------------------------------------------------------------------------------------------------------------------------

In [1]:
from sqlalchemy import create_engine, text
import pandas as pd

#### setting up engine to use sql querries

In [2]:
database_name = 'scooters'

connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

In [3]:
engine = create_engine(connection_string)

### Checking to see if companies are in compliance with 'removal of trips below one minute'

In [4]:
trip_query = '''
SELECT *
FROM trips;
'''

with engine.connect() as connection:
    trips = pd.read_sql(text(trip_query), con = connection)

In [5]:
trips.loc[trips.companyname == 'Bolt Mobility', 'tripduration'] = trips.tripduration / 60

In [21]:
trips_below_1min = trips[(trips['tripduration'] < 1)] 

In [22]:
trips_below_1min

pubtimestamp companyname triprecordnum          sumdid  \
749    2019-05-04 17:20:16.330        Bird       BRD1073  PoweredUNKNOWN   
853    2019-05-04 17:40:17.463        Bird       BRD1236  PoweredUNKNOWN   
854    2019-05-04 17:40:17.463        Bird       BRD1235  PoweredUNKNOWN   
913    2019-05-04 17:53:14.590        Lyft       LFT1239   Powered403246   
994    2019-05-04 18:05:18.530        Bird       BRD1505  PoweredUNKNOWN   
...                        ...         ...           ...             ...   
565231 2019-05-04 16:47:54.167        Lyft       LFT1115   Powered837779   
565252 2019-05-04 16:53:45.933        Lyft       LFT1128   Powered585413   
565310 2019-05-04 17:05:14.947        Bird        BRD927  PoweredUNKNOWN   
565328 2019-05-04 17:10:02.037        Lyft       LFT1151   Powered055940   
565354 2019-05-04 17:11:31.463        Lyft       LFT1159   Powered444815   

        tripduration  tripdistance   startdate        starttime     enddate  \
749         0.000000           0.0  2019-05-04  17:20:33.226666  2019-05-04   
853         0.000000           0.0  2019-05-04  17:41:44.116666  2019-05-04   
854         0.000000           0.0  2019-05-04  17:42:04.193333  2019-05-04   
913         0.843350           0.0  2019-05-04  17:52:23.896666  2019-05-04   
994         0.000000           0.0  2019-05-04  18:05:08.866666  2019-05-04   
...              ...           ...         ...              ...         ...   
565231      0.319183           0.0  2019-05-04  16:47:34.843333  2019-05-04   
565252      0.312617           0.0  2019-05-04  16:53:27.080000  2019-05-04   
565310      0.000000           0.0  2019-05-04  17:05:51.610000  2019-05-04   
565328      0.617917           0.0  2019-05-04  17:09:24.856666  2019-05-04   
565354      0.640783           0.0  2019-05-04  17:10:52.906666  2019-05-04   

                endtime  startlatitude  startlongitude  endlatitude  \
749     17:20:43.316666       36.15180       -86.78420     36.15180   
853     17:41:53.580000       36.15830       -86.77730     36.15830   
854     17:42:15.456666       36.16020       -86.77850     36.16020   
913     17:53:14.496666       36.15170       -86.78414     36.15167   
994     18:05:32.160000       36.16470       -86.77610     36.16470   
...                 ...            ...             ...          ...   
565231  16:47:53.993333       36.15555       -86.78671     36.15554   
565252  16:53:45.836666       36.16063       -86.78194     36.16066   
565310  17:06:16.783333       36.16210       -86.77630     36.16210   
565328  17:10:01.933333       36.15844       -86.77690     36.15845   
565354  17:11:31.353333       36.15639       -86.78071     36.15639   

        endlongitude                                          triproute  \
749        -86.78420                                                 []   
853        -86.77730                                                 []   
854        -86.77850                                                 []   
913        -86.78421  [(36.1517, -86.78414), (36.15168, -86.7842), (...   
994        -86.77610                                                 []   
...              ...                                                ...   
565231     -86.78671     [(36.15555, -86.78671), (36.15554, -86.78671)]   
565252     -86.78190      [(36.16063, -86.78194), (36.16066, -86.7819)]   
565310     -86.77630                                                 []   
565328     -86.77690       [(36.15844, -86.7769), (36.15845, -86.7769)]   
565354     -86.78076     [(36.15639, -86.78071), (36.15639, -86.78076)]   

                     create_dt  
749    2019-05-05 05:31:14.100  
853    2019-05-05 05:31:17.403  
854    2019-05-05 05:31:17.370  
913    2019-05-05 08:34:28.753  
994    2019-05-05 05:31:21.310  
...                        ...  
565231 2019-05-05 08:34:24.480  
565252 2019-05-05 08:34:24.863  
565310 2019-05-05 05:31:12.377  
565328 2019-05-05 08:34:25.533  
565354 2019-05-05 08:34:25.783  

[9424

In [23]:
#seeing how many entries are non-compliant with rule 2 per each company

trips_below_1min.groupby(['companyname'])['sumdid'].count()

companyname
Bird             3963
Bolt Mobility     270
Lime              661
Lyft             4530
Name: sumdid, dtype: int64

### Amount of entries that are non-compliant with rule 2

#### Bird: 3,963
#### Gotcha: 270
#### Lime: 661
#### Lyft: 4,530

### Checking to see if companies are in compliance with 'trip lengths are capped at 24 hours'

In [24]:
trips_above_24hour = trips[(trips['tripduration'] > (24 * 60))]

In [25]:
trips_above_24hour

pubtimestamp    companyname triprecordnum  \
11841  2019-05-05 19:08:31.353           Lyft       LFT1105   
80057  2019-05-17 01:23:46.463           SPIN        SPI283   
94314  2019-05-18 14:20:06.423           SPIN         SPI43   
117937 2019-05-20 22:02:08.773           SPIN        SPI304   
124551 2019-05-22 03:23:14.603           SPIN        SPI254   
133884 2019-05-23 23:14:58.743           SPIN        SPI192   
143172 2019-05-24 23:50:53.553           SPIN        SPI450   
150476 2019-05-25 18:16:36.020           SPIN         SPI17   
179996 2019-05-27 19:18:22.330           SPIN        SPI733   
186295 2019-05-28 16:14:06.130           SPIN       SPI1229   
197311 2019-05-30 21:06:01.163           SPIN        SPI122   
198254 2019-05-30 23:41:06.850           SPIN       SPI1308   
201386 2019-05-31 14:11:40.230           SPIN         SPI10   
204691 2019-05-31 21:16:56.820           SPIN        SPI812   
206418 2019-06-01 19:42:46.580           SPIN        SPI291   
206721 2019-06-13 05:17:41.630           SPIN        SPI146   
235803 2019-06-04 00:34:41.057           SPIN       SPI1042   
301741 2019-06-13 03:27:38.330           SPIN        SPI145   
301774 2019-06-13 03:32:38.447           SPIN         SPI82   
302350 2019-06-13 05:57:42.743           SPIN         SPI70   
302363 2019-06-13 06:02:42.787           SPIN        SPI763   
302364 2019-06-13 06:02:42.787           SPIN        SPI525   
302368 2019-06-13 06:07:43.103           SPIN        SPI104   
336457 2019-06-17 02:41:03.780           SPIN        SPI564   
349245 2019-06-19 22:18:31.667           SPIN        SPI115   
383422 2019-06-26 08:33:59.480           Lyft        LFT428   
454810 2019-07-10 01:05:36.000  Bolt Mobility      BOL00008   
464325 2019-07-13 04:13:31.000  Bolt Mobility      BOL00146   
465937 2019-07-13 05:34:32.417           SPIN        SPI171   
471204 2019-07-16 21:26:59.000  Bolt Mobility      BOL00198   
477555 2019-07-14 22:05:24.000  Bolt Mobility      BOL00284   
508073 2019-07-22 01:52:24.023           SPIN         SPI42   
508075 2019-07-22 01:52:24.023           SPIN         SPI43   
541326 2019-07-30 07:23:04.170           SPIN        SPI433   

                                              sumdid  tripduration  \
11841                                  Powered413161   1662.259483   
80057                                 Powered2251745   6133.000000   
94314                                 Powered5839762   6496.000000   
117937                                Powered4567425   1923.000000   
124551                                Powered6905839   1759.000000   
133884                                Powered2327092   1442.000000   
143172                                Powered2043594   9000.000000   
150476                                Powered6197635  11156.000000   
179996                                Powered7605539   2987.000000   
186295                                Powered9505409   2507.000000   
197311                                Powered9423528   4014.000000   
198254                                Powered1932731   7332.000000   
201386                                Powered3047571   2271.000000   
204691                                Powered5173676   5670.000000   
206418                                Powered6661538   5791.000000   
206721                                Powered5750098  36206.000000   
235803                                Powered1943283   1542.000000   
301741                                Powered7331723   2221.000000   
301774                                Powered2882478   1441.000000   
302350                                Powered6843540   7500.000000   
302363                                Powered3522267   4885.000000   
302364                                Powered1440517   5007.000000   
302368                                Powered4590795   3075.000000   
336457                                Powered9080472   3269.000000   
349245                                Powered6903219   1443.00

In [26]:
#seeing how many entries are non-compliant with rule 3 per each company

trips_above_24hour.groupby(['companyname'])['sumdid'].count()

companyname
Bolt Mobility     4
Lyft              2
SPIN             28
Name: sumdid, dtype: int64

### Amount of entries that are non-compliant with rule 3

#### Bolt Mobility:4
#### Lyft: 2
#### SPIN: 28

### Checking to see which companies were in compliance with both rules

In [12]:
compliant_query = '''
SELECT DISTINCT(companyname)
FROM trips
WHERE companyname NOT IN ('Bird', 'Gotcha', 'Lime', 'Lyft', 'Bolt Mobility', 'SPIN');
'''

with engine.connect() as connection:
    compliant_companies = pd.read_sql(text(compliant_query), con = connection)

In [13]:
compliant_companies

companyname
0        JUMP

### JUMP is the only company that is compliant with rules 2 and 3

# The code below will join the tables and bring in only the data that meets the 'cleaned' requirements.

In [14]:
cleaning_query = '''
SELECT *
       
FROM trips
INNER JOIN scooters
ON trips.sumdid = scooters.sumdid
    AND trips.pubtimestamp = scooters.pubdatetime
    
WHERE tripduration < (24 * 60)
    AND tripduration > 1
LIMIT 5;
'''

with engine.connect() as connection:
    cleaned_df = pd.read_sql(text(cleaning_query), con = connection)

In [15]:
cleaned_df[['sumdid','companyname','chargelevel','latitude', 'longitude', 'startlatitude', 'startlongitude','endlatitude', 'endlongitude','pubtimestamp','pubdatetime', 'create_dt', 'startdate', 'starttime']]

sumdid                sumdid companyname companyname  \
0  Powered7IM4P5IZ3SC6T  Powered7IM4P5IZ3SC6T        Lime        Lime   
1  PoweredAS6DJPP44XMZX  PoweredAS6DJPP44XMZX        Lime        Lime   
2  PoweredKJBLE2LZLQ4KF  PoweredKJBLE2LZLQ4KF        Lime        Lime   
3  Powered3IA7BYM7HIZYZ  Powered3IA7BYM7HIZYZ        Lime        Lime   
4  Powered4HFURN2XVEO4F  Powered4HFURN2XVEO4F        Lime        Lime   

   chargelevel   latitude  longitude  startlatitude  startlongitude  \
0         60.0  36.150442 -86.781600      36.160280      -86.775370   
1         58.0  36.140694 -86.793208      36.157499      -86.788756   
2         98.0  36.154188 -86.785508      36.176916      -86.750524   
3         80.0  36.160905 -86.778706      36.159784      -86.772896   
4          9.0  36.157418 -86.774894      36.153063      -86.789612   

   endlatitude  endlongitude            pubtimestamp             pubdatetime  \
0    36.150266    -86.781084 2019-07-07 22:55:56.247 2019-07-07 22:55:56.247   
1    36.153031    -86.783720 2019-07-07 22:55:56.247 2019-07-07 22:55:56.247   
2    36.158553    -86.771569 2019-05-04 21:18:30.000 2019-05-04 21:18:30.000   
3    36.160905    -86.778706 2019-05-08 20:27:13.267 2019-05-08 20:27:13.267   
4    36.155603    -86.783649 2019-05-08 20:27:13.270 2019-05-08 20:27:13.270   

                create_dt   startdate starttime  
0 2019-07-08 09:01:39.810  2019-07-07  22:08:02  
1 2019-07-08 09:01:39.843  2019-07-07  22:24:40  
2 2019-05-05 07:51:13.317  2019-05-03  23:56:45  
3 2019-05-09 09:05:37.637  2019-05-08  20:21:21  
4 2019-05-09 09:05:37.927  2019-05-08  20:18:56